### PR TITLE
refator reorder

### PR DESCRIPTION
**Pull request by NachmanBerkowitz**

### CODE REVIEW

File: projects/table-builder/src/lib/classes/table-store.ts [hunk 1]
```diff
@@ -180,7 +193,48 @@ export class TableStore extends ComponentStore<TableState> {
...
+      userDefinedOrderArr.forEach(([key,order])=>userDefinedOrder[key]=order);
+    }
+    return userDefinedOrder;
+    
+    function thereIsSavedOrder(){
+      return !!userDefinedOrderArr.length;
+    }
+
+    function viewableMetaAddedSinceLastSave(){
+      return userDefinedOrderArr.length < viewableMetaDataArr.length
```
Comment (e-davidson) [inline projects/table-builder/src/lib/classes/table-store.ts:222]: you're not taking into account 1 item added and 1 item removed.

File: projects/table-builder/src/lib/classes/table-store.ts [hunk 2]
```diff
@@ -180,7 +193,48 @@ export class TableStore extends ComponentStore<TableState> {
...
+    const order = this.initializeOrder(state, metaData);
+    return {...state, metaData, sorted, userDefined:{...state.userDefined,order:order}};
   });
 
+  private initializeOrder = (state:TableState,mds: Dictionary<MetaData>) : Dictionary<number> => {
+    const viewableMetaDataArr = Object.values(mds).sort((a,b)=>a.order-b.order).filter(a => a.fieldType !== FieldType.Hidden);
+    const userDefinedOrder = state.userDefined.order;
+    let userDefinedOrderArr = Object.entries(userDefinedOrder);
+
+    if( thereIsSavedOrder() && viewableMetaAddedSinceLastSave()){
```
Comment (e-davidson) [inline projects/table-builder/src/lib/classes/table-store.ts:204]: in my opinion.
In a case where the developer added a column just remove the user defined array.
Like I said, there's no good way to know what the user meant.
He could mean column 2 he could mean before column 3 or after column 1.

**Review by e-davidson — CHANGES_REQUESTED** [review]

File: projects/table-builder/src/lib/classes/table-store.ts [hunk 3]
```diff
@@ -103,13 +103,26 @@ export class TableStore extends ComponentStore<TableState> {
...
   });
 
+  setUserDefinedOrder = this.updater((state,moved:{newOrder:number,oldOrder:number})=>{
+    const userDefinedOrder = {...state.userDefined.order};
+
+    const {newOrder ,oldOrder} = moved;
+    const mdsArr = orderViewableMetaData(state);
+    moveItemInArray(mdsArr,oldOrder,newOrder);
+    const mds = {...state.metaData};
+    mdsArr.forEach((md,index) => {
```
[the remark this answers is not in the document]
Comment (NachmanBerkowitz) [reply projects/table-builder/src/lib/classes/table-store.ts]: done

File: projects/table-builder/src/lib/classes/table-store.ts [hunk 4]
```diff
@@ -180,7 +193,48 @@ export class TableStore extends ComponentStore<TableState> {
     if(sorted.length === 0) {
       sorted = this.createPreSort(metaData);
     }
-    return {...state, metaData, sorted};
+    const order = this.initializeOrder(state, metaData);
+    return {...state, metaData, sorted, userDefined:{...state.userDefined,order:order}};
   });
 
+  private initializeOrder = (state:TableState,mds: Dictionary<MetaData>) : Dictionary<number> => {
+    const viewableMetaDataArr = Object.values(mds).sort((a,b)=>a.order-b.order).filter(a => a.fieldType !== FieldType.Hidden);
```
[the remark this answers is not in the document]
Comment (NachmanBerkowitz) [reply projects/table-builder/src/lib/classes/table-store.ts]: not nogea anymore

**Review by e-davidson — APPROVED** [review]

**Pull request merged**